File: apps/android/feature/insights/ui/src/main/kotlin/com/scribblefit/feature/insights/ui/components/VolumeChart.kt
package com.scribblefit.feature.insights.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.scribblefit.core.designsystem.ScribbleFitTheme
import com.scribblefit.feature.insights.domain.model.VolumeDataPoint
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

@Composable
internal fun VolumeChart(
    points: List<VolumeDataPoint>,
    modifier: Modifier = Modifier,
) {
    if (points.isEmpty()) return

    val density = LocalDensity.current
    val lineColor = ScribbleFitTheme.colors.primary
    val fillColorTop = ScribbleFitTheme.colors.primary.copy(alpha = 0.12f)
    val fillColorBottom = ScribbleFitTheme.colors.primary.copy(alpha = 0.0f)
    val dotColor = ScribbleFitTheme.colors.primary
    val haloColor = ScribbleFitTheme.colors.surfaceContainerLow // Match background for contrast
    val labelColor = ScribbleFitTheme.colors.midGray

    val textMeasurer = rememberTextMeasurer()
    val labelStyle = TextStyle(fontSize = 10.sp, color = labelColor)
    val dateFormatter = remember { DateTimeFormatter.ofPattern("MMM d") }

    val maxVolume = points.maxOf { it.volume }
    val minVolume = points.minOf { it.volume }
    val midVolume = (maxVolume + minVolume) / 2f
    val range = (maxVolume - minVolume).coerceAtLeast(1f) // Ensure no division by zero

    fun formatAxisValue(value: Float): String {
        return when {
            value >= 1_000_000 -> String.format("%.1fM", value / 1_000_000f)
            value >= 1000 -> String.format("%.1fk", value / 1000f)
            else -> String.format("%.0f", value)
        }
    }

    Surface(
        color = ScribbleFitTheme.colors.surfaceContainerLow,
        shape = RoundedCornerShape(ScribbleFitTheme.shapes.medium),
        modifier = modifier,
    ) {
        Column(modifier = Modifier.padding(ScribbleFitTheme.spacing.medium)) {
            Canvas(modifier = Modifier.fillMaxSize()) {
                val yAxisWidth = with(density) { 48.dp.toPx() } // Slightly wider for large numbers
                val xAxisHeight = with(density) { 24.dp.toPx() }
                val dotRadius = with(density) { 4.5.dp.toPx() }
                val haloRadius = with(density) { 6.5.dp.toPx() }
                val strokeWidth = with(density) { 2.5.dp.toPx() }

                val chartLeft = yAxisWidth
                val chartRight = size.width - with(density) { 8.dp.toPx() } // Right margin for dot
                val chartTop = with(density) { 12.dp.toPx() } // Top margin for dot
                val chartBottom = size.height - xAxisHeight
                val chartWidth = (chartRight - chartLeft).coerceAtLeast(1f)
                val chartHeight = (chartBottom - chartTop).coerceAtLeast(1f)

                // Y-axis labels
                val yLabels = if (maxVolume > minVolume) {
                    listOf(
                        maxVolume to chartTop,
                        midVolume to chartTop + chartHeight / 2f,
                        minVolume to chartBottom
                    )
                } else {
                    listOf(maxVolume to chartTop + chartHeight / 2f)
                }

                for ((value, y) in yLabels) {
                    val text = formatAxisValue(value)
                    val measured = textMeasurer.measure(text, labelStyle)
                    drawText(
                        textLayoutResult = measured,
                        topLeft = Offset(
                            x = (yAxisWidth - measured.size.width - with(density) { 12.dp.toPx() }).coerceAtLeast(
                                0f
                            ),
                            y = y - measured.size.height / 2f
                        )
                    )
                }

                fun pointOffset(index: Int): Offset {
                    val x = if (points.size > 1) {
                        chartLeft + (index.toFloat() / (points.size - 1)) * chartWidth
                    } else {
                        chartLeft + chartWidth / 2f
                    }

                    val normalizedY = if (maxVolume > minVolume) {
                        (points[index].volume - minVolume) / range
                    } else {
                        0.5f
                    }
                    val y = chartBottom - normalizedY * chartHeight
                    return Offset(x, y)
                }

                if (points.size >= 2) {
                    val linePath = Path().apply {
                        val first = pointOffset(0)
                        moveTo(first.x, first.y)
                        for (i in 1 until points.size) {
                            val prev = pointOffset(i - 1)
                            val curr = pointOffset(i)
                            val cpX1 = prev.x + (curr.x - prev.x) / 2f
                            val cpX2 = prev.x + (curr.x - prev.x) / 2f
                            cubicTo(cpX1, prev.y, cpX2, curr.y, curr.x, curr.y)
                        }
                    }

                    // Fill
                    val fillPath = Path().apply {
                        addPath(linePath)
                        lineTo(pointOffset(points.size - 1).x, chartBottom)
                        lineTo(pointOffset(0).x, chartBottom)
                        close()
                    }
                    drawPath(
                        path = fillPath,
                        brush = Brush.verticalGradient(
                            colors = listOf(fillColorTop, fillColorBottom),
                            startY = chartTop,
                            endY = chartBottom
                        )
                    )

                    // Line
                    drawPath(
                        path = linePath,
                        color = lineColor,
                        style = Stroke(
                            width = strokeWidth,
                            cap = StrokeCap.Round,
                            join = StrokeJoin.Round
                        )
                    )
                }

                // Dots with halo
                for (i in points.indices) {
                    val p = pointOffset(i)
                    drawCircle(color = haloColor, radius = haloRadius, center = p)
                    drawCircle(color = dotColor, radius = dotRadius, center = p)
                }

                // X-axis date labels
                val xLabelIndices = when {
                    points.size <= 3 -> points.indices.toList()
                    else -> listOf(0, points.size / 2, points.size - 1)
                }.distinct()

                for (i in xLabelIndices) {
                    val date = Instant.ofEpochMilli(points[i].date)
                        .atZone(ZoneId.systemDefault())
                        .toLocalDate()
                    val text = dateFormatter.format(date)
                    val measured = textMeasurer.measure(text, labelStyle)
                    val p = pointOffset(i)
                    val labelX = (p.x - measured.size.width / 2f)
                        .coerceIn(
                            chartLeft - measured.size.width / 2f,
                            size.width - measured.size.width
                        )
                    drawText(
                        textLayoutResult = measured,
                        topLeft = Offset(labelX, chartBottom + with(density) { 6.dp.toPx() })
                    )
                }
            }
        }
    }
}

File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureInsights/Sources/FeatureInsights/UI/Components/VolumeChart.swift
import SwiftUI
import CoreModel
import CoreDesignSystem

struct VolumeChart: View {
    let points: [VolumeDataPoint]
    let weightUnitLabel: String

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        return formatter
    }()

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            GeometryReader { _ in
                Canvas { context, size in
                    if points.isEmpty { return }

                    let maxVolume = points.map { $0.volume }.max() ?? 0
                    let minVolume = points.map { $0.volume }.min() ?? 0
                    let range = max(maxVolume - minVolume, 1)

                    let yAxisWidth: CGFloat = 40
                    let xAxisHeight: CGFloat = 20
                    let dotRadius: CGFloat = 4.5
                    let haloRadius: CGFloat = 6.5
                    let strokeWidth: CGFloat = 2.5

                    let chartLeft = yAxisWidth
                    let chartRight = size.width - 8
                    let chartTop: CGFloat = 12
                    let chartBottom = size.height - xAxisHeight
                    let chartWidth = max(chartRight - chartLeft, 1)
                    let chartHeight = max(chartBottom - chartTop, 1)

                    func pointOffset(_ index: Int) -> CGPoint {
                        let x = points.count > 1
                            ? chartLeft + CGFloat(index) / CGFloat(points.count - 1) * chartWidth
                            : chartLeft + chartWidth / 2

                        let normalizedY = maxVolume > minVolume
                            ? CGFloat((points[index].volume - minVolume) / range)
                            : 0.5

                        let y = chartBottom - normalizedY * chartHeight
                        return CGPoint(x: x, y: y)
                    }

                    // Draw Y-axis labels (Max, Mid, Min)
                    let labelFont = Font.system(size: 10)
                    let midVolume = (maxVolume + minVolume) / 2
                    let yLabels = maxVolume > minVolume
                        ? [(maxVolume, chartTop), (midVolume, chartTop + chartHeight / 2), (minVolume, chartBottom)]
                        : [(maxVolume, chartTop + chartHeight / 2)]

                    for (value, y) in yLabels {
                        let text = Text(formatAxisValue(value))
                            .font(labelFont)
                            .foregroundColor(.secondary)

                        context.draw(text, at: CGPoint(x: yAxisWidth / 2, y: y), anchor: .center)
                    }

                    if points.count >= 2 {
                        var linePath = Path()
                        let first = pointOffset(0)
                        linePath.move(to: first)

                        for i in 1..<points.count {
                            let prev = pointOffset(i - 1)
                            let curr = pointOffset(i)
                            let cp1 = CGPoint(x: prev.x + (curr.x - prev.x) / 2, y: prev.y)
                            let cp2 = CGPoint(x: prev.x + (curr.x - prev.x) / 2, y: curr.y)
                            linePath.addCurve(to: curr, control1: cp1, control2: cp2)
                        }

                        // Draw Gradient Fill
                        var fillPath = linePath
                        fillPath.addLine(to: CGPoint(x: pointOffset(points.count - 1).x, y: chartBottom))
                        fillPath.addLine(to: CGPoint(x: pointOffset(0).x, y: chartBottom))
                        fillPath.closeSubpath()

                        context.fill(
                            fillPath,
                            with: .linearGradient(
                                Gradient(colors: [Color.scribblePrimary.opacity(0.12), Color.scribblePrimary.opacity(0)]),
                                startPoint: CGPoint(x: 0, y: chartTop),
                                endPoint: CGPoint(x: 0, y: chartBottom)
                            )
                        )

                        // Draw Line
                        context.stroke(
                            linePath,
                            with: .color(Color.scribblePrimary),
                            style: StrokeStyle(lineWidth: strokeWidth, lineCap: .round, lineJoin: .round)
                        )
                    }

                    // Draw Dots with Halo
                    for i in 0..<points.count {
                        let p = pointOffset(i)
                        context.fill(
                            Path(ellipseIn: CGRect(x: p.x - haloRadius, y: p.y - haloRadius, width: haloRadius * 2, height: haloRadius * 2)),
                            with: .color(Color.scribbleSurfaceContainerLow)
                        )
                        context.fill(
                            Path(ellipseIn: CGRect(x: p.x - dotRadius, y: p.y - dotRadius, width: dotRadius * 2, height: dotRadius * 2)),
                            with: .color(Color.scribblePrimary)
                        )
                    }

                    // Draw X-axis labels (Start, Mid, End)
                    let xLabelIndices = points.count <= 3
                        ? Array(0..<points.count)
                        : [0, points.count / 2, points.count - 1]

                    let dateFormatter = VolumeChart.dateFormatter

                    for i in xLabelIndices {
                        let p = pointOffset(i)
                        let text = Text(dateFormatter.string(from: points[i].date))
                            .font(labelFont)
                            .foregroundColor(.secondary)

                        context.draw(text, at: CGPoint(x: p.x, y: chartBottom + 12), anchor: .top)
                    }
                }
            }
        }
        .frame(height: 180)
        .padding(ScribbleFitSpacing.medium)
        .background(Color.scribbleSurfaceContainerLow)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .accessibilityIdentifier("volumeChart")
    }

    private func formatAxisValue(_ value: Float) -> String {
        if value >= 1_000_000 {
            return String(format: "%.1fM", value / 1_000_000)
        }
        if value >= 1000 {
            return String(format: "%.1fk", value / 1000)
        }
        return String(format: "%.0f", value)
    }
}
